### PR TITLE
Improve performance of filters (2-3 x faster)

### PR DESCRIPTION
**Pull request by snakefoot**
ExpressionFilters - Compile MethodInfo into LateBoundMethod-delegate
===

Using the technique from Nate Kohari:

https://web.archive.org/web/20140226131612/http://kohari.org/2009/03/06/fast-late-bound-invocation-with-expression-trees

Then the MethodInfo is compiled into a delegate, that is much faster. Manual handling of default-/optional-parameters are required.

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1779)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Impressive!

**Comment by 304NotModified** [comment]
Any idea how much faster?

**Comment by snakefoot** [comment]
| Test | Time | Msgs/sec | GC2 | GC1 | GC0 |
------|------|-----------|------|-----|-------|
| No Filter | 2565ms | 779518 | 39 | 50 | 57 |
| Filter | 9789ms | 204291 | 68 | 93 | 117 |
| LateBound | 3693ms | 541445 | 36 | 54 | 61 |

**Comment by 304NotModified** [comment]
\0/

File: src/NLog/Internal/ReflectionHelpers.cs [hunk 1]
```diff
@@ -92,6 +92,72 @@ public static bool IsStaticClass(this Type type)
...
+                    parameterType = parameterType.GetElementType();
+
+                var valueCast = Expression.Convert(valueObj, parameterType);
+
+                parameterExpressions.Add(valueCast);
+            }
+
+            // non-instance for static method, or ((TInstance)instance)
+            var instanceCast = methodInfo.IsStatic ? null :
+                Expression.Convert(instanceParameter, methodInfo.ReflectedType);
```
Comment (304NotModified) [inline src/NLog/Internal/ReflectionHelpers.cs]: @snakefoot 

another netstandard challenge, `methodInfo.ReflectedType` isn't there. Any idea how to fix it? I dunno what the `ReflectedType` is in this case.

Comment (snakefoot) [inline src/NLog/Internal/ReflectionHelpers.cs]: Seems that we should be using DeclaringType instead of ReflectedType

**Pull request merged**